li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java

```diff
@@ -55,14 +61,22 @@
   private UUID _clientId;
 
   // Per cluster Consumers
-  private Map<ClusterDescriptor, LiKafkaConsumer<K, V>> _consumers;
+  private ConcurrentSkipListMap<ClusterDescriptor, LiKafkaConsumer<K, V>> _consumers;
```

Comment:
this is theoretically correct, although im concerned about the runtime overhead of the eadMap and tailMap below.
may i suggest a volatile array of {descriptor, consumer} tuples? realistically, for <10 clusters an array would be faster and we could just index into it.
concurrency could be handled by atomically swapping the array for a new one (so the poll()ing thread only pays for a single volatile read)

Reply:
Done. This class is not thread-safe and I didn't really check for concurrent updates.